client/src/contexts/UndoRedoContext.tsx

```diff
@@ -1,103 +1,249 @@
-      if (!state) {
-        return;
-      }
-      canvas.loadFromJSON(state, () => {
-        canvas.renderAll();
-        canvas.calcOffset();
-      });
+    if (!canvas || redoStack.current.length === 0) {
+      return;
+    }
+    const lastAction = redoStack.current.pop();
+    if (!lastAction) {
+      return;
     }
+
+    lastAction.redo();
+    undoStack.current.push(lastAction);
+    canvas.renderAll();
   }, []);
 
   useEffect(() => {
-    if (canvas) {
-      const handleMouseDown = () => {
-        saveState();
-      };
-
-      const handleMouseUp = () => {
-        saveState();
-      };
-
-      const handlePathCreated = () => {
-        saveState();
-      };
-
-      canvas.on("mouse:down", handleMouseDown);
-      canvas.on("mouse:up", handleMouseUp);
-      canvas.on("path:created", handlePathCreated);
-
-      return () => {
-        canvas.off("mouse:down", handleMouseDown);
-        canvas.off("mouse:up", handleMouseUp);
-        canvas.off("path:created", handlePathCreated);
-      };
+    if (!canvas) {
+      return;
     }
-  }, [canvas, saveState]);
-  return <UndoRedoContext.Provider value={{ saveState, undo, redo }}>{children}</UndoRedoContext.Provider>;
+
+    const handlePathCreated = (e: any) => {
+      if (!e.path) {
+        return;
+      }
+
+      if (!canvas.contains(e.path)) {
+        return;
+      } // to differentiate betwen 'real paths' and 'eraser paths'
+
+      const id = generateId("path");
+      Object.assign(e.path, { id });
+
+      const command = new AddCommand(canvas, e.path.toJSON(["id"]));
+      saveCommand(command);
+    };
+
+    const handleActiveSelectionModification = () => {
+      const activeObjects: fabric.Object[] = canvas.getActiveObjects();
+
+      if (activeObjects.length <= 1) {
+        return;
+      }
+
+      const activeObjectsJSONs = activeObjects.map((obj: fabric.Object) => getJsonWithAbsoluteProperties(obj));
+
+      const commands: UndoRedoCommand[] = [];
+      for (const i in activeObjectsJSONs) {
+        const activeObjectJSON = activeObjectsJSONs[i];
+        const recentlyActiveObjectJSON = recentlyActiveObjects[i];
+
+        commands.push(new ModifyCommand(canvas, recentlyActiveObjectJSON, activeObjectJSON, handleStyleChange));
+      }
+      saveCommand(new ComplexCommand(commands));
+    };
+
+    /**
+     * This handler handles oly those modifications, which are visible in the `fabric.IEvent#transform.original` object
+     * (like angle, scaleX, top, left etc.) (not like color, strokeWidth etc.)
+     */
+    const handleObjectModified = (e: fabric.IEvent) => {
+      if (e.target?.type === "activeSelection") {
+        handleActiveSelectionModification();
+        return;
+      }
+
+      const oldValues = e.transform?.original; // the original properties values before modification
+      const modifiedObject = e.target;
+
+      if (!oldValues || !modifiedObject) {
+        return;
+      }
+
+      modifiedObject.clone(
```

Comment:
Yea, we discussed the `clone(...)` method, and you suggested changing it to something like this:

```js
const objectJSON = fabricObject.toJSON(["id"]);
const clonedObject = Object.clone(objectJSON);
```

If you're able to get that approach working, that would be great! But if it doesn’t work, no worries.